bftengine/src/bftengine/messages/ReplicaRestartReadyMsg.cpp

```diff
@@ -0,0 +1,64 @@
+// Concord
+//
+// Copyright (c) 2020 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").  You may not use this product except in
+// compliance with the Apache 2.0 License.
+//
+// This product may include a number of subcomponents with separate copyright notices and license terms. Your use of
+// these subcomponents is subject to the terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+#include <cstring>
+#include "OpenTracing.hpp"
+#include "ReplicaRestartReadyMsg.hpp"
+#include "SysConsts.hpp"
+#include "Crypto.hpp"
+#include "SigManager.hpp"
+
+namespace bftEngine {
+namespace impl {
+
+ReplicaRestartReadyMsg::ReplicaRestartReadyMsg(ReplicaId srcReplicaId,
+                                               SeqNum s,
+                                               uint16_t sigLen,
+                                               const concordUtils::SpanContext& spanContext)
+    : MessageBase(srcReplicaId, MsgCode::ReplicaRestartReady, spanContext.data().size(), sizeof(Header) + sigLen) {
+  b()->genReplicaId = srcReplicaId;
+  b()->seqNum = s;
+  b()->sigLength = sigLen;
+  std::memcpy(body() + sizeof(Header), spanContext.data().data(), spanContext.data().size());
+}
+
+ReplicaRestartReadyMsg* ReplicaRestartReadyMsg::create(ReplicaId senderId,
+                                                       SeqNum s,
+                                                       const concordUtils::SpanContext& spanContext) {
+  auto sigManager = SigManager::instance();
+  const size_t sigLen = sigManager->getMySigLength();
+
+  ReplicaRestartReadyMsg* m = new ReplicaRestartReadyMsg(senderId, s, sigLen, spanContext);
+
+  auto position = m->body() + sizeof(Header);
+  std::memcpy(position, spanContext.data().data(), spanContext.data().size());
+  position += spanContext.data().size();
+
+  sigManager->sign(m->body(), sizeof(Header), position, sigLen);
+
+  return m;
+}
+
+void ReplicaRestartReadyMsg::validate(const ReplicasInfo& repInfo) const {
+  auto sigManager = SigManager::instance();
+  auto totalSize = sizeof(Header) + spanContextSize();
+  if (size() < totalSize || !repInfo.isIdOfReplica(idOfGeneratedReplica()))
+    throw std::runtime_error(__PRETTY_FUNCTION__ + std::string(": basic validations"));
```

Comment:
It is ok to throw an exception but make sure that someone is catching it in some centralized place where you call all validate functions, and drop the message if there is additional processing to be done.
if such a place exists already...then this code is fine.
same comment for below code.

Reply:
Yes this exception is caught, logged and dropped in ReplicaImp::validateMessage()